### PR TITLE
SC-617: allow drivers, automations and zones to update their own config

### PR DESCRIPTION
**Pull request by hexaglow**
In preparation for user-editable config, drivers (etc.) need a way to trigger their own configuration to be modified.

Provide a new `service.ConfigUpdater` to each driver/auto/zone when constructing it. 
The service may use it to change its own configuration. The new config will be applied (using `service.Lifecycle.Configure(...)`), and if successful, persisted into the app's active config.

Also modify the implementation of `ServicesApi.ConfigureService` to store the updated configurations in the app's active config, instead of the old overlay directories. 

This isn't quite complete yet: we should have an example and/or tests of using the new functionality.
Also, `ServicesApi.CreateService` doesn't result in a runnable service - the app needs to restart first.

### CODE REVIEW

File: pkg/app/appconf/store.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+		if typ != "" && typ != existing.Type {
+			return errors.New("driver type cannot be changed")
+		}
+		driverCfg = existing
+		driverCfg.Raw = data
+		updated.Drivers[idx] = driverCfg
+	}
+
+	// test that the updated config marshalls successfully - we don't want to store a bad config
+	_, err := json.Marshal(updated)
```
Comment (mattnathan) [inline pkg/app/appconf/store.go:121]: Could we get away with just testing the new/updated driver config marshals? I'm afraid that if we add bulk update features later this json.Marshal will cause it to be O(n!) complexity

File: pkg/app/appconf/store.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+	}
+	s.active = updated
+	return nil
+}
+
+type DriverStore struct {
+	store *Store
+}
+
+func (ds *DriverStore) SaveConfig(_ context.Context, name string, typ string, data []byte) error {
```
Comment (mattnathan) [inline pkg/app/appconf/store.go:82]: It is very annoying that these type stores can't be more generic

File: pkg/app/services.go [hunk 3]
```diff
@@ -224,3 +228,22 @@ func announceSystemServices[M ~map[string]T, T any](c *Controller, services *ser
...
+
+func (s *serviceConfigStore) UpdateConfig(ctx context.Context, data []byte) error {
+	serv := s.services.Get(s.id)
+	if serv == nil {
+		return fmt.Errorf("service %s not found", s.id)
+	}
+	_, err := serv.Service.Configure(data)
+	if err != nil {
+		return err
+	}
```
Comment (mattnathan) [inline pkg/app/services.go:246]: I'm not 100% sold on this bit. 

I think as the driver/etc will be the one calling this method, it already knows its config is updating and should handle it internally. I also think we might get issues here with ctx being cancelled before Save is called, it entirely depends on whether a driver/etc links the ctx given during `applyConfig` to the one passed to UpdateConfig.

If I'm thinking about it I think there are two separate concerns going on here:
1. The system wants to tell the driver the config has updated
2. The driver wants to tell the system the config has updated

Number 1 is for the general case: boot, generic APIs, getting config from the cloud, etc.
Number 2 is for the runtime case: submitting forms, discovery (of devices), user editing, etc.

File: pkg/task/serviceapi/api.go [hunk 4]
```diff
@@ -568,4 +538,12 @@ func (a *Api) storeConfig(ctx context.Context, id, kind string, data []byte) err
...
 
-var fileNameEncoder = base32.StdEncoding.WithPadding(base32.NoPadding)
+type Store interface {
+	// SaveConfig will persist the configuration for a service.
+	// The data must be an encoded JSON object.
+	// name identifies the service, and is required.
+	// typ is the type of the service. If the store already contains a service with the given name, typ may be empty,
+	// in which case the existing type is used. If the store does not contain a service with the given name, typ must
+	// be non-empty.
+	SaveConfig(ctx context.Context, name string, typ string, data []byte) error
```
Comment (mattnathan) [inline pkg/task/serviceapi/api.go:548]: `name string, typ string` -> `name, typ string`

**Review by mattnathan — CHANGES_REQUESTED** [review]

File: pkg/app/appconf/store.go (hunk 2 above)
Comment (hexaglow) [reply pkg/app/appconf/store.go:82]: I'll have a go at it, but my gut feeling is that it won't reduce the line count much

File: pkg/app/appconf/store.go (hunk 1 above)
Comment (hexaglow) [reply pkg/app/appconf/store.go]: Yes, good idea. Done.

File: pkg/app/services.go (hunk 3 above)
Comment (hexaglow) [reply pkg/app/services.go:246]: I see what you mean. It's more flexible if we let the service handle reconfiguring itself. 

I'll try to make this clear in the documentation.

Comment (hexaglow) [reply pkg/app/services.go]: I've change the semantics of `ConfigUpdater` to not call `Configure()` any more.

**Review by mattnathan — APPROVED** [review]

**Comment by hexaglow** [comment]
> Also, `ServicesApi.CreateService` doesn't result in a runnable service - the app needs to restart first.

Going to leave this as a future ticket because it's a nice-to-have rather than strictly needed.

**Pull request merged**